### PR TITLE
fix: clone async checkpoint tensors to CPU to prevent GPU OOM

### LINKED ISSUE #21630: AsyncCheckpointIO Should Clone() to CPU not on GPU
### Bug description

In [pytorch-lightning/src/lightning/pytorch/plugins/io/async_plugin.py](https://github.com/Lightning-AI/pytorch-lightning/blame/master/src/lightning/pytorch/plugins/io/async_plugin.py#L100)

Tensor in Model Checkpoint are cloned() to prevent race condition when doing async uploading in threads.

```
# snapshot the checkpoint payload on the caller thread to avoid races with parameter mutation
def _clone_tensor(t: torch.Tensor) -> torch.Tensor:
    """Clones a tensor on the caller thread."""
    # detach to avoid autograd history and clone to take a point-in-time copy
    return t.detach().clone()
```

However, here it is clone() from GPU memory to GPU memory. Given GPU memory is often limited, this step is dangerous when num_thread go up. 

A more clean solution is to clone the tensors to CPU. This achieves the same purpose without using without using GPU memory. CPU memory is abundant most of the time. 

```
# snapshot the checkpoint payload on the caller thread to avoid races with parameter mutation
def _clone_tensor(t: torch.Tensor) -> torch.Tensor:
    """Clones a tensor on the caller thread."""
    # detach to avoid autograd history and clone to take a point-in-time copy
    return t.detach().cpu().clone()
```

### What version are you seeing the problem on?

master

### Reproduced in studio

_No response_

### How to reproduce the bug

```python

```

### Error messages and logs

If cloning on GPU memory, this is dangerous for large model checkpoint (e.g. 15GB in our case)

Cloning to CPU 
```
[ASYNC CHECKPOINT BEFORE clone] GPU 0: allocated=21.54 GB, reserved=124.30 GB
[ASYNC CHECKPOINT AFTER clone] GPU 0: allocated=21.54 GB, reserved=124.30 GB
```

Cloning on GPU
```
[ASYNC CHECKPOINT BEFORE clone] GPU 0: allocated=21.54 GB, reserved=124.30 GB
[ASYNC CHECKPOINT AFTER clone] GPU 0: allocated=37.54 GB, reserved=124.30 GB
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- PyTorch Lightning Version (e.g., 2.6.0):
#- PyTorch Version (e.g., 2.5):
#- Python version (e.g., 3.12):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
```

</details>

### More info

_No response_

cc @ethanwharris

### PR DESCRIPTION
**Pull request by karthik-idikuda**
## Summary

Clone async checkpoint tensors to CPU to prevent GPU OOM during async saves.

## Details

`AsyncCheckpointIO._clone_tensor()` previously called `t.detach().clone()`, which allocates new GPU memory for each cloned tensor. For large model checkpoints (e.g., 15GB+), this **doubles GPU memory usage** during checkpoint saves, which can cause OOM errors.

### Before (GPU clone):
```
[ASYNC CHECKPOINT BEFORE clone] GPU 0: allocated=21.54 GB
[ASYNC CHECKPOINT AFTER clone]  GPU 0: allocated=37.54 GB  ← +16 GB!
```

### After (CPU clone):
```
[ASYNC CHECKPOINT BEFORE clone] GPU 0: allocated=21.54 GB
[ASYNC CHECKPOINT AFTER clone]  GPU 0: allocated=21.54 GB  ← no change
```

### Changes
- **`async_plugin.py`**: Changed `t.detach().clone()` → `t.detach().cpu().clone()` in `_clone_tensor()`. Moving to CPU first avoids doubling GPU memory. CPU memory is typically abundant and this still prevents the race condition with parameter mutation.
- **`test_async_checkpoint.py`**: Added `test_async_checkpoint_clones_tensors_to_cpu()` to verify cloned tensors are on CPU and retain correct values.

Fixes #21630

----
📚 Documentation preview 📚: https://pytorch-lightning--21631.org.readthedocs.build/en/21631/

### CODE REVIEW

File: tests/tests_pytorch/plugins/test_async_checkpoint.py [hunk 1]
```diff
@@ -51,3 +51,19 @@ def test_async_checkpoint_should_snapshot_values_before_mutation():
         "AsyncCheckpointIO must snapshot the checkpoint (clone tensors) on the main thread "
         "to avoid races with parameter mutation; got mutated value instead"
     )
+
+
+@pytest.mark.filterwarnings("ignore::DeprecationWarning")
+def test_async_checkpoint_clones_tensors_to_cpu():
```
Comment (deependujha) [inline tests/tests_pytorch/plugins/test_async_checkpoint.py]: This test will run on GitHub CI, with no GPUs, and only CPU. I don't think it's sufficient.

File: tests/tests_pytorch/plugins/test_async_checkpoint.py [hunk 2]
```diff
@@ -51,3 +51,19 @@ def test_async_checkpoint_should_snapshot_values_before_mutation():
         "AsyncCheckpointIO must snapshot the checkpoint (clone tensors) on the main thread "
         "to avoid races with parameter mutation; got mutated value instead"
     )
+
+
+@pytest.mark.filterwarnings("ignore::DeprecationWarning")
```
Comment (deependujha) [inline tests/tests_pytorch/plugins/test_async_checkpoint.py]: why do we need this?

File: src/lightning/pytorch/plugins/io/async_plugin.py [hunk 3]
```diff
@@ -95,6 +95,8 @@ def teardown(self) -> None:
...
 # snapshot the checkpoint payload on the caller thread to avoid races with parameter mutation
 def _clone_tensor(t: torch.Tensor) -> torch.Tensor:
-    """Clones a tensor on the caller thread."""
-    # detach to avoid autograd history and clone to take a point-in-time copy
-    return t.detach().clone()
+    """Clones a tensor to CPU on the caller thread."""
+    # detach to avoid autograd history, move to CPU to avoid doubling GPU memory usage, and clone to take a
+    # point-in-time copy. Moving to CPU first is important because clone() on a CUDA tensor allocates new GPU memory,
+    # which can cause OOM errors for large model checkpoints.
+    return t.detach().cpu().clone()
```
Comment (TheGreatFrankie) [inline src/lightning/pytorch/plugins/io/async_plugin.py]: For GPU tensors, .cpu() already creates a copy when moving from GPU to CPU, making the subsequent .clone() redundant and wasteful (doubles memory copy). For correctness this works, but performance suffers. For CPU tensors, .cpu() is a no-op so .clone() is necessary. Consider: t.clone().detach().cpu() if t.is_cuda else t.detach().clone() for better efficiency.
```
def _clone_tensor_to_cpu(t: torch.Tensor) -> torch.Tensor:
    """Clone a tensor to CPU memory.

    Detaches from autograd, moves to CPU, and clones to ensure the checkpoint
    holds a point-in-time snapshot that won't be mutated by ongoing training.

    For GPU tensors, `cpu()` already allocates a new CPU copy so `clone()` is
    skipped. For CPU tensors, `cpu()` is a no-op so `clone()` is required.
    """
    return t.detach().cpu() if t.is_cuda else t.detach().clone()
```

File: tests/tests_pytorch/plugins/test_async_checkpoint.py (hunk 1 above)
Comment (karthik-idikuda) [reply tests/tests_pytorch/plugins/test_async_checkpoint.py]: You're right that this test runs CPU-only on CI. It still validates the core contract: `_clone_tensor` must produce an independent CPU copy whose storage is not shared with the original. The GPU-specific path (`t.is_cuda` branch) avoids the redundant `clone()` but can only be tested on GPU hardware — the CPU path exercises the `clone()` fallback which is the more subtle correctness requirement.

File: tests/tests_pytorch/plugins/test_async_checkpoint.py (hunk 2 above)
Comment (karthik-idikuda) [reply tests/tests_pytorch/plugins/test_async_checkpoint.py]: This test verifies that `_clone_tensor` actually produces a tensor on CPU with independent storage (i.e., mutating the original does not affect the clone). Without this, a future refactor could accidentally break the snapshot guarantee and the only signal would be silent data corruption in saved checkpoints.

File: src/lightning/pytorch/plugins/io/async_plugin.py (hunk 3 above)
Comment (karthik-idikuda) [reply src/lightning/pytorch/plugins/io/async_plugin.py]: Great catch — updated. For CUDA tensors `cpu()` already allocates a new host-memory copy, so the extra `clone()` was redundant. The function now branches on `t.is_cuda`: GPU tensors use `detach().cpu()`, CPU tensors use `detach().clone()`. Pushed in b55dbc089.

File: tests/tests_pytorch/plugins/test_async_checkpoint.py (hunk 2 above)
Comment (deependujha) [reply tests/tests_pytorch/plugins/test_async_checkpoint.py]: I mean for using `deprecation warning` suppression.

File: src/lightning/pytorch/plugins/io/async_plugin.py (hunk 3 above)
Comment (deependujha) [reply src/lightning/pytorch/plugins/io/async_plugin.py]: I believe `cpu` is a special case, not `cuda`, as we also support tpus (not sure about amd).

so better approach imo will be: `return t.detach().clone() if t.is_cpu else t.detach().cpu()`

what do you think @TheGreatFrankie?

File: src/lightning/pytorch/plugins/io/async_plugin.py [hunk 4]
```diff
@@ -95,6 +95,16 @@ def teardown(self) -> None:
...
+    that won't be mutated by ongoing training.
+
+    For CUDA tensors ``cpu()`` already allocates a new host-memory copy, so an
+    extra ``clone()`` is unnecessary.  For CPU tensors ``cpu()`` is a no-op, so
+    ``clone()`` is required to break storage sharing.
+
+    """
+    if t.is_cuda:
+        return t.detach().cpu()
     return t.detach().clone()
```
Comment (deependujha) [inline src/lightning/pytorch/plugins/io/async_plugin.py]:
```suggestion
    if t.is_cpu:
        return t.detach().clone()
    return t.detach().cpu()
```

**Comment by justusschock** [comment]
I am not sure we want this to change. The reason being that a transfer from GPU to CPU is a synchronization point and the whole point of async checkpointing is to avoid those. If in doubt, you can still use synchronous checkpointing.

File: src/lightning/pytorch/plugins/io/async_plugin.py (hunk 3 above)
Comment (TheGreatFrankie) [reply src/lightning/pytorch/plugins/io/async_plugin.py]: @deependujha Good catch 
I think we should do cpu against all other
I just confirmed there is no is_cpu field for tensor. But we can do
```
t = t.detach()
return t.clone() if t.device.type == "cpu" else t.cpu()
```

**Comment by TheGreatFrankie** [comment]
> I am not sure we want this to change. The reason being that a transfer from GPU to CPU is a synchronization point and the whole point of async checkpointing is to avoid those. If in doubt, you can still use synchronous checkpointing.

🤔 but the original GPU to GPU clone in memory is also a synchronization point. I don't fully understand your point. 
If GPU to GPU clone is a synchronization point and GPU to CPU clone is also a synchronization point. why don't we clone to CPU to save GPU memory?

File: src/lightning/pytorch/plugins/io/async_plugin.py (hunk 3 above)
Comment (deependujha) [reply src/lightning/pytorch/plugins/io/async_plugin.py]: interesting, this works for me lol.
```python
>>> import torch
>>> t = torch.tensor(1.0)
>>> t
tensor(1.)
>>> t.is_cpu
True
>>> t.is_cuda
False
```

Comment (TheGreatFrankie) [reply src/lightning/pytorch/plugins/io/async_plugin.py]: Intersting, seems like is_cpu is not on [the PyTorch official doc](https://docs.pytorch.org/docs/stable/generated/torch.Tensor.is_cuda.html) (Here I only find is_cuda). I didn't find it in the [raw code for tensor class](https://github.com/pytorch/pytorch/blob/main/torch/_tensor.py?utm_source=chatgpt.com) either

Maybe it exist as a internal property. But anyway this is a nit thing. We can do either.

File: src/lightning/pytorch/plugins/io/async_plugin.py (hunk 4 above)
Comment (karthik-idikuda) [reply src/lightning/pytorch/plugins/io/async_plugin.py]: Good point — updated. CPU is the special case that needs `clone()`, while all other devices (CUDA, TPU, XLA) produce a new copy via `cpu()`. Now using `t.is_cpu` to branch — it's available as a public property since PyTorch 1.x. Pushed in 5e61027.
